### ISSUE TITLE
dtype signature for arithmetic is not fixed

### ISSUE THREAD

**Issue by shoyer** [numpy/numpy#6240]
NumPy operations generally follow the rule that the output dtype/shape is a deterministic function of the inputs dtype/shape.

However, binary arithmetic with 0d arrays appears to be an exception to these rules. Consider:

```
np.array(12).dtype # int64
np.array(4625196817309499392).dtype # int64
(np.array([[123]], dtype=np.float32) + np.array(12)).dtype # float32
(np.array([[123]], dtype=np.float32) + np.array(4625196817309499392)).dtype # float64
```

Basically, whether the result is float32 depends on the size of the scalar argument.

I can see why this would be desirable for operations with Python scalars (e.g., `array + 1`), but when the 0d/scalar argument is already typed this is highly surprising.

---

EDIT seberg:  This is basically what NEP 50 is about.

**Comment by seberg** [comment]
I agree.... Right now there is no difference between a 0-d array or a scalar (you could also think of it as: all scalars get converted to arrays always).
That means that if you want special behaviour for scalars, you also have the special behaviour for 0-d arrays. One of those things that should be improved but are deep down and likely hard to impossible to change.

**Comment by njsmith** [comment]
Yeah, that is weird and I don't think it's even documented. The special
rule that we ignore the bitwidth of scalars when doing promotion is
documented, but I don't know what's going on with the value sensitivity. I
think a few things like this showed up in 1.6 or so and since then we've
removed some but not all of them?
.
One idea that came up at SciPy is that when we're overhauling casting as
part of the dtype overhaul, maybe we will want to remove the special casing
for 0d objects, and instead have special casing for the built-in python
scalars only, where they get mapped to unspevified-width dtypes that have
special permissive casting rules. Obviously this is just a vague idea at
this point though...

**Comment by mhvk** [comment]
@njsmith - +1 for not special-casing 0-d arrays. As it is a bit off-topic, perhaps a separate issue?

**Comment by seberg** [comment]
This was changed by NEP 50 and thus fixed in main/NumPy 2.0.
